FE310-based boards: transition from pinmux to pinctrl

fkokosinski opened the pull request:
This pull request contains the following changes as part of the effort in #39740:
* introduction of the new pinctrl driver for FE310-based boards
* modifications of the existing drivers used in FE310-based boards so that they use the pinctrl API
* binding changes
* removal of the obsolete pinmux counterpart

fkokosinski committed:
drivers: pinctrl: add SiFive pinctrl driver

Add a pinctrl driver used in FE310-based boards.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: hifive1: add pinctrl configuration for hifive1 board

Add pinctrl configuration for future use with the new pinctrl driver.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: hifive1_revb: add pinctrl configuration for hifive1_revb board

Add pinctrl configuration for future use with the new pinctrl driver.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: qemu_riscv32: add pinctrl configuration for qemu_riscv32_xip

Add pinctrl configuration for future use with the new pinctrl driver.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: FE310-based boards: transition to pinctrl driver

This commit makes the transition from the pinmux driver to the pinctrl
driver. It also modifies UART, SPI and I2C drivers used in FE310-based
boards to use the new pinctrl API.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
drivers: pinmux: remove SiFive pinmux driver

After transitioning to the new pinctrl driver, it is now obsolete.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

gmarull commented on drivers/pinctrl/Kconfig.sifive:
```diff
@@ -0,0 +1,8 @@
+
+config PINCTRL_SIFIVE
+	bool "SiFive Freedom SoC pinmux driver"
+	depends on SOC_SERIES_RISCV_SIFIVE_FREEDOM
```
please default based on `dt_compat_enabled` (see others)

fkokosinski replied:
Done

gmarull commented on boards/riscv/hifive1/hifive1-pinctrl.dtsi:
```diff
@@ -0,0 +1,79 @@
+
+&pinctrl {
+	/* UART0 */
+	uart0_rx_default: uart0_rx_default {
+		pinmux = <16 SIFIVE_PINMUX_IOF0>;
+	};
+	uart0_tx_default: uart0_tx_default {
+		pinmux = <17 SIFIVE_PINMUX_IOF0>;
+	};
```
suggestion: if not autogenerating nodes, grouping representation leads to a less verbose board pinctrl file.

gmarull commented on boards/riscv/hifive1/hifive1_defconfig:
```diff
@@ -10,7 +10,8 @@ CONFIG_UART_SIFIVE_PORT_0=y
 CONFIG_PINMUX=y
-CONFIG_PINMUX_SIFIVE=y
+CONFIG_PINCTRL=y
+CONFIG_PINCTRL_SIFIVE=y
```
only `CONFIG_PINMUX=y` should be required, see my previous comment about defaulting based on dt_compat_status. If all boards need this flag, consider enabling this at SoC level.

fkokosinski replied:
Removed `CONFIG_PINCTRL_SIFIVE=y`

gmarull commented on dts/bindings/spi/sifive,spi0.yaml:
```diff
@@ -5,8 +5,14 @@ description: Sifive SPI controller
     reg:
       required: true
+
+    pinctrl-0:
+      required: false
+
+    pinctrl-names:
+      required: false
```
this is redundant, none are required by default. applies to others as well.

fkokosinski replied:
Removed redundant fields from binding yamls

gmarull commented on boards/riscv/hifive1/hifive1_defconfig:
```diff
@@ -10,7 +10,8 @@ CONFIG_UART_SIFIVE_PORT_0=y
 CONFIG_UART_CONSOLE=y
 CONFIG_PLIC=y
 CONFIG_PINMUX=y
```
pinmux driver seems to be removed, so it should be this option

fkokosinski replied:
Done.

gmarull reviewed (changes requested):
Thanks for the contribution, I have a few comments (some non-blocking)

gmarull commented on dts/bindings/pinctrl/sifive,pinctrl.yaml:
```diff
@@ -0,0 +1,27 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  SiFive FE310 IO Function (iof)
+  Binding covers the IOF_EN/IOF_SEL registers that are a subset of the GPIO
+  controller.
```
nitpick: please extend this documentation on how pinctrl is expected to be used (see other vendors as a reference)

fkokosinski replied:
Done.

fkokosinski committed:
soc: riscv: sifive-freedom: fix SYS_CLOCK_HW_CYCLES_PER_SEC value

This commit fixes the default value of SYS_CLOCK_HW_CYCLES_PER_SEC
option. The previous value of 32768 is not consistent with the
documentation of FE310 SoC. Only FE310-based boards rely on the default
value of this option; other boards from the Freedom series define it
themselves.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

gmarull commented on drivers/pinctrl/pinctrl_sifive.c:
```diff
@@ -0,0 +1,52 @@
+
+	/* Enable IO function for this pin */
+	val = sys_read32(PINCTRL_IOF_EN);
+	val |= (1 << pin);
```
```suggestion
	val |= BIT(pin);
```

fkokosinski replied:
Done.

gmarull commented on drivers/serial/uart_sifive.c:
```diff
@@ -334,6 +339,13 @@ static int uart_sifive_init(const struct device *dev)
 	cfg->cfg_func();
 #endif
 
+#ifdef CONFIG_PINCTRL
+	ret = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+	if (ret < 0) {
+		return ret;
+	}
+#endif
```
does driver work at all without pinctrl? why is this made optional (same applies to others)?

fkokosinski replied:
This change was necessary for the `hifive_unleashed` and `hifive_unmatched` boards, as they weren't using the old pinmux driver before.

fkokosinski replied:
They don't do it at all, if I'm correct. I've re-checked this with their SoCs manuals (FU540 and FU740 respectively), and the section that would describe IOFs is empty.
